pytorch_lightning/metrics/classification/precision_recall.py

```diff
@@ -11,205 +11,322 @@
-        num_classes: int = 1,
+        num_classes: Optional[int] = None,
         threshold: float = 0.5,
-        average: str = 'micro',
+        average: str = "micro",
         multilabel: bool = False,
+        mdmc_average: Optional[str] = None,
+        ignore_index: Optional[int] = None,
+        top_k: Optional[int] = None,
+        is_multiclass: Optional[bool] = None,
         compute_on_step: bool = True,
         dist_sync_on_step: bool = False,
         process_group: Optional[Any] = None,
+        dist_sync_fn: Callable = None,
     ):
+        allowed_average = ["micro", "macro", "weighted", "samples", "none", None]
+        if average not in allowed_average:
+            raise ValueError(f"The `average` has to be one of {allowed_average}, got {average}.")
+
         super().__init__(
+            reduce="macro" if average in ["weighted", "none", None] else average,
+            mdmc_reduce=mdmc_average,
+            threshold=threshold,
+            top_k=top_k,
+            num_classes=num_classes,
+            is_multiclass=is_multiclass,
+            ignore_index=ignore_index,
             compute_on_step=compute_on_step,
             dist_sync_on_step=dist_sync_on_step,
             process_group=process_group,
+            dist_sync_fn=dist_sync_fn,
         )
 
-        self.num_classes = num_classes
-        self.threshold = threshold
         self.average = average
-        self.multilabel = multilabel
-
-        assert self.average in ('micro', 'macro'), \
-            "average passed to the function must be either `micro` or `macro`"
 
-        self.add_state("true_positives", default=torch.zeros(num_classes), dist_reduce_fx="sum")
-        self.add_state("predicted_positives", default=torch.zeros(num_classes), dist_reduce_fx="sum")
+    def compute(self) -> torch.Tensor:
+        """
+        Computes the precision score based on inputs passed in to ``update`` previously.
 
-    def update(self, preds: torch.Tensor, target: torch.Tensor):
-        preds, target = _input_format_classification_one_hot(
-            self.num_classes, preds, target, self.threshold, self.multilabel
-        )
+        Return:
+            The shape of the returned tensor depends on the ``average`` parameter
 
-        # multiply because we are counting (1, 1) pair for true positives
-        self.true_positives += torch.sum(preds * target, dim=1)
-        self.predicted_positives += torch.sum(preds, dim=1)
+            - If ``average in ['micro', 'macro', 'weighted', 'samples']``, a one-element tensor will be returned
+            - If ``average in ['none', None]``, the shape will be ``(C,)``, where ``C`` stands  for the number
+              of classes
+        """
+        if isinstance(self.tp, list):
+            tp = torch.cat(self.tp)
+            fp = torch.cat(self.fp)
+            tn = torch.cat(self.tn)
+            fn = torch.cat(self.fn)
+        else:
+            tp, fp, tn, fn = self.tp, self.fp, self.tn, self.fn
```

Comment:
should this be made into a class method as it seems to be done in every class?
```suggestion
        tp, fp, tn, fn = self._get_final_stats()
```

Reply:
That's a good point, will do that